### PR TITLE
feature: add new api request to get protocol params

### PR DESCRIPTION
**Pull request by SorinC6**
task: https://emurgo.atlassian.net/browse/YOMO-986

### CODE REVIEW

File: packages/api/src/app/api/protocol-params.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+export const isFrontendProtocolParamsResponse = createTypeGuardFromSchema(
+  AppFrontendFeesResponseSchema,
+)
+
+export const parseFrontendProtocolParamsResponse = (
+  data: App.FrontendProtocolParamsResponse,
+): App.FrontendProtocolParamsResponse | undefined => {
+  const filteredObject = pick(data, [
+    'epoch',
```
Comment (SorinC6) [inline packages/api/src/app/api/protocol-params.ts]: have filtered some of the data as the endpoint response is quite big. I'm thinking that we can return only the constants that we need from the api package

File: apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts [hunk 2]
```diff
@@ -602,6 +602,9 @@ export const makeShelleyWallet = (constants: typeof MAINNET | typeof TESTNET) =>
       const changeAddr = await this.getAddressedChangeAddress()
       const addressedUtxos = await this.getAddressedUtxos()
 
+      const protocolParams = await this.getProtocolParams()
+      const {coins_per_utxo_word, key_deposit, min_utxo, pool_deposit, min_fee_a, min_fee_b} = protocolParams
+
```
Comment (SorinC6) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts]: not entirely sure what are all the constants that we will replace and get from the api

File: apps/wallet-mobile/src/yoroi-wallets/hooks/index.ts [hunk 3]
```diff
@@ -630,6 +631,29 @@ export const useFrontendFees = (
...
+    ...options,
+    queryFn: () => wallet.api.getProtocolParams(),
+  })
+
+  return {
+    ...query,
+    protocolParams: query.data,
+  }
+}
+
```
Comment (SorinC6) [inline apps/wallet-mobile/src/yoroi-wallets/hooks/index.ts:645]: Have added this hook as well but not sure we will actually use 🤔

File: apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts [hunk 4]
```diff
@@ -158,9 +155,13 @@ export const makeShelleyWallet = (constants: typeof MAINNET | typeof TESTNET) =>
   }
 
   const appApi = AppApi.appApiMaker({baseUrl: API_ROOT})
+  const cardanoApi = CardanoApi.cardanoApiMaker({baseUrl: API_ROOT_NEW})
 
   return class ShelleyWallet implements YoroiWallet {
     readonly api: App.Api = appApi
+
+    readonly cardanoApi: CardanoApi.api = cardanoApi
+
```
Comment (banklesss) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts]:
```suggestion
    readonly cardanoApi: CardanoApi.api = cardanoApi
```

File: packages/api/src/index.ts [hunk 5]
```diff
@@ -59,39 +64,48 @@ export const CardanoApi = {
...
+  export interface NftRecords extends ApiNftRecords {}
+
+  export type MetadataFile = ApiMetadataFile
+  export type TokenId = ApiTokenId
+
+  export type ProtocolParamsResult = ApiProtocolParamsResult
+
+  export interface api {
+    getProtocolParams: () => Promise<ProtocolParamsResult>
   }
```
Comment (banklesss) [inline packages/api/src/index.ts:110]: I don't like this but I need it. Open to ideas.

File: packages/api/src/index.ts [hunk 6]
```diff
@@ -59,39 +64,48 @@ export const CardanoApi = {
...
+  export type NftMetadataRecord = ApiNftMetadataRecord
+  export interface NftRecords extends ApiNftRecords {}
+
+  export type MetadataFile = ApiMetadataFile
+  export type TokenId = ApiTokenId
+
+  export type ProtocolParamsResult = ApiProtocolParamsResult
+
+  export interface api {
+    getProtocolParams: () => Promise<ProtocolParamsResult>
```
Comment (banklesss) [inline packages/api/src/index.ts]: I don't like this but I need it. Open to ideas.

File: apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts [hunk 7]
```diff
@@ -158,9 +155,11 @@ export const makeShelleyWallet = (constants: typeof MAINNET | typeof TESTNET) =>
   }
 
   const appApi = AppApi.appApiMaker({baseUrl: API_ROOT})
+  const cardanoApi = CardanoApi.cardanoApiMaker({baseUrl: API_ROOT_NEW})
```
Comment (banklesss) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts]: Are the rest of cardano apis working with the new api root?

File: apps/wallet-mobile/src/yoroi-wallets/cardano/byron/ByronWallet.ts [hunk 8]
```diff
@@ -127,9 +128,11 @@ const networkId = NETWORK_REGISTRY.HASKELL_SHELLEY
 const implementationId = WALLET_IMPLEMENTATION_REGISTRY.HASKELL_BYRON
 
 const appApi = AppApi.appApiMaker({baseUrl: API_ROOT})
+const cardanoApi = CardanoApi.cardanoApiMaker({baseUrl: API_ROOT_NEW})
 
 export class ByronWallet implements YoroiWallet {
   readonly api: App.Api = appApi
+  readonly cardanoApi: CardanoApi.api = cardanoApi
```
Comment (banklesss) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/byron/ByronWallet.ts]: comment here https://github.com/Emurgo/yoroi/pull/2918/files#r1424340689

File: apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts (hunk 7 above)
Comment (SorinC6) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts]: @banklesss the new api root is used only for the `protocol params`. Later when the new backend is ready we will use only this `new api` and remove the old one

Comment (SorinC6) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts]: - yoroi-backend: the one Yoroi always used. Legacy, bad and to be completely replaced by yoroi-backend-zero;

- yoroi-backend-zero: new backend that will replace yoroi-backend, but right now you will use only the /protocolparameters endpoint of it;

Comment (banklesss) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts]: we have to consider that if we want to extend cardanoApi

cc @stackchain

File: apps/wallet-mobile/src/yoroi-wallets/cardano/constants/mainnet/constants.ts [hunk 9]
```diff
@@ -39,9 +39,11 @@ export const BASE_CONFIG = {
 
 export const IS_MAINNET = true
 export const API_ROOT = 'https://api.yoroiwallet.com/api'
+export const API_ROOT_NEW = 'https://yoroi-backend-zero-mainnet.emurgornd.com'
```
Comment (banklesss) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/constants/mainnet/constants.ts:42]: @SorinC6 do you know if its the same api root for mainnet and preprod?

Comment (banklesss) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/constants/mainnet/constants.ts:42]: sorry. i noticed already

File: apps/wallet-mobile/src/yoroi-wallets/hooks/index.ts [hunk 10]
```diff
@@ -630,6 +631,28 @@ export const useFrontendFees = (
...
+  options?: UseQueryOptions<
+    CardanoApi.ProtocolParamsResult,
+    Error,
+    CardanoApi.ProtocolParamsResult,
+    ['protocol-params']
+  >,
+) => {
+  const query = useQuery({
+    suspense: true,
+    queryKey: ['protocol-params'],
```
Comment (michalszorad) [inline apps/wallet-mobile/src/yoroi-wallets/hooks/index.ts]: Maybe `[walletId, 'protocol-params']` as each wallet can have different protocol params depending on the network, right?

File: apps/wallet-mobile/src/yoroi-wallets/cardano/byron/ByronWallet.ts [hunk 11]
```diff
@@ -697,23 +700,28 @@ export class ByronWallet implements YoroiWallet {
...
-            constant: containsDatum
-              ? String(BigInt(networkConfig.LINEAR_FEE.CONSTANT) * 2n)
-              : networkConfig.LINEAR_FEE.CONSTANT,
+            coefficient,
+            constant: containsDatum ? String(BigInt(constant) * 2n) : constant,
           },
           minimumUtxoVal: networkConfig.MINIMUM_UTXO_VAL,
-          coinsPerUtxoWord: networkConfig.COINS_PER_UTXO_WORD,
-          poolDeposit: networkConfig.POOL_DEPOSIT,
+          coinsPerUtxoWord: String(Number(coinsPerUtxoByte) * 8),
```
Comment (stackchain) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/byron/ByronWallet.ts:733]: is it needed the `* 8`?

File: apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts (hunk 7 above)
Comment (stackchain) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/shelley/ShelleyWallet.ts]: the apiMaker should receive the `network` only, and the endpoints should be part of the module, not the wallet.

File: apps/wallet-mobile/src/yoroi-wallets/cardano/constants/mainnet/constants.ts (hunk 9 above)
Comment (stackchain) [inline apps/wallet-mobile/src/yoroi-wallets/cardano/constants/mainnet/constants.ts]: move to the package as API_ENDPOINTS[network][endpoint] ie. governance

File: packages/api/src/cardano/api/cardano-api-maker.ts [hunk 12]
```diff
@@ -0,0 +1,17 @@
+import {Fetcher, fetcher} from '@yoroi/common'
+
+import {getProtocolParams as getProtocolParamsWrapper} from './protocol-params'
+import {CardanoApi} from '../../index'
+export const cardanoApiMaker = ({
+  baseUrl,
```
Comment (stackchain) [inline packages/api/src/cardano/api/cardano-api-maker.ts]: @banklesss update to receive the network `mainnet` | `preprod` | `sanchonet`, also create a `API_ENDPOINTS` config, i.e -> governance.

File: packages/api/src/cardano/api/token-offchain-metadata.ts [hunk 13]
```diff
@@ -6,24 +6,18 @@ import {
...
-  ApiOffChainMetadataRecord,
-  ApiOffChainMetadataRequest,
-  ApiOffChainMetadataResponse,
-  ApiTokenId,
-  ApiTokenRegistryEntry,
-  ApiTokenRegistryProperty,
-} from './types'
+import {ApiTokenRegistryProperty} from './types'
 import {asSubject} from '../translators/transformers/asSubject'
+import {CardanoApi} from '../../index'
```
Comment (stackchain) [inline packages/api/src/cardano/api/token-offchain-metadata.ts]: don't import from `/index` it usually leads to circular dips

File: packages/api/src/index.ts (hunk 6 above)
Comment (stackchain) [reply packages/api/src/index.ts]: if its supposed to represent the module -> `CardanoApi.Module` -> export interface `Module` -> so the factory will be returning this object.

**Review by stackchain — CHANGES_REQUESTED** [review]
@banklesss if you could prioritize it.

File: apps/wallet-mobile/src/yoroi-wallets/cardano/byron/ByronWallet.ts (hunk 11 above)
Comment (banklesss) [reply apps/wallet-mobile/src/yoroi-wallets/cardano/byron/ByronWallet.ts:733]: params API gives `coinsPerUtxoByte`. I assume `createUnsignedTx` etc will accept `coinsPerUtxoByte` in the near future right?

**Review by michalszorad — APPROVED** [review]

**Pull request merged**